lib/burndown_chart/versions_helper_patch.rb

```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+require_dependency 'versions_helper'
+
+module BurndownChart
+  module VersionsHelperPatch
+    def self.included(base)
+      base.send(:prepend, InstanceMethods)
+    end
+
+    module InstanceMethods
+      def issues_burndown_chart_data(version)
+        return nil if version.visible_fixed_issues.empty?
+        chart_start_date = (version.start_date || version.created_on).to_date
+        chart_end_date = (version.due_date || (version.completed? ? version.visible_fixed_issues.maximum(:closed_on) : User.current.today)).to_date
+        line_end_date = [User.current.today, chart_end_date].min
+        step_size = ((chart_start_date..chart_end_date).count.to_f / 90).ceil
+        issues = version.visible_fixed_issues
+        return nil if step_size < 1
+
+        ideal = [{ :t => chart_start_date.to_s, :y => issues.count }, { :t => version.due_date.to_s, :y => 0 }]
+        total_closed = chart_start_date.step(line_end_date, step_size).collect{ |d| { :t => d.to_s, :y => issues.where("#{Issue.table_name}.closed_on IS NULL OR #{Issue.table_name}.closed_on>=?", d.end_of_day).count } }
+        open = chart_start_date.step(line_end_date, step_size).collect{ |d| { :t => d.to_s, :y => issues.where("#{Issue.table_name}.created_on<=?", d.end_of_day).count - issues.open(false).where("#{Issue.table_name}.closed_on<=?", d.end_of_day).count } }
```

Comment:
total_closedとopenについて、
* line_start_dateが2021/3/8
* line_end_dateがUser.current.today(2021/3/15)
* stepが5

という状態で2021/3/15時点でチケットをクローズしたり追加した場合、それがチャートに反映されないことが分かりました。

これは、`chart_start_date.step(line_end_date, step_size)`が上のような例では `[Mon, 08 Mar 2021, Sat, 13 Mar 2021]` のようなEnumeratorを返してしまうため、2021/3/13〜2021/3/15の期間中の件数がチェックされないためだと思います。  
もう少し良いやり方もあるかもしれませんが、次のsuggestionのように `line_end_date + (step_size - 1)` を基準に変えることで、件数をチェックする範囲が[Mon, 08 Mar 2021, Sat, 13 Mar 2021, Thu, 18 Mar 2021]に変わって2021/3/15の変更もチャートに反映されるようになります。  
(ただ、そう変更した場合User.current.todayより先の日付の場所に点が入ることになるので、それも少し違和感につながりそうではあります)

```suggestion
        total_closed = chart_start_date.step(line_end_date + (step_size - 1), step_size).collect{ |d| { :t => d.to_s, :y => issues.where("#{Issue.table_name}.closed_on IS NULL OR #{Issue.table_name}.closed_on>=?", d.end_of_day).count } }
        open = chart_start_date.step(line_end_date + (step_size - 1), step_size).collect{ |d| { :t => d.to_s, :y => issues.where("#{Issue.table_name}.created_on<=?", d.end_of_day).count - issues.open(false).where("#{Issue.table_name}.closed_on<=?", d.end_of_day).count } }
```

Reply:
step_sizeを設けることでチャート中のドットが増えすぎるのを回避しているのですが、最新(User.current.today)のチケット数がプロットされないのは誤解を生む原因になりそうですね 😓 
一方で、提案のコードのように User.current.today より先の日付にドットを置くのも避けたい...
そこで、以下のように、line_end_dateが必ずデータに含まれるようにして対処しようと考えています。
```suggestion
        total_closed = (chart_start_date.step(line_end_date, step_size).to_a + [line_end_date]).uniq.collect{ |d| { :t => d.to_s, :y => issues.where("#{Issue.table_name}.closed_on IS NULL OR #{Issue.table_name}.closed_on>=?", d.end_of_day).count } }
        open = (chart_start_date.step(line_end_date, step_size).to_a + [line_end_date]).uniq.collect{ |d| { :t => d.to_s, :y => issues.where("#{Issue.table_name}.created_on<=?", d.end_of_day).count - issues.open(false).where("#{Issue.table_name}.closed_on<=?", d.end_of_day).count } }
```
感想をお聞かせください。

Reply:
コミット d3fb9a9ffe8bb4647279f08fef09e1254b36aee8 にて、プロットする日付にラインの終了日が常に含まれるように変更しました。